File: ctci/3_4.cpp
#include <iostream>
#include <stack>
#include <stdexcept>

class MyQueue {
   private:
    std::stack<int> in_stack;
    std::stack<int> out_stack;

    void in_to_out() {
        while (!in_stack.empty()) {
            out_stack.push(in_stack.top());
            in_stack.pop();
        }
    }

   public:
    bool is_empty() const {
        return in_stack.empty() && out_stack.empty();
    }

    void push(int x) {
        in_stack.push(x);
    }

    int peek() {
        if (is_empty()) {
            throw std::runtime_error("Queue is empty");
        }
        if (out_stack.empty()) in_to_out();
        return out_stack.top();
    }

    int pop() {
        if (is_empty()) {
            throw std::runtime_error("Queue is empty");
        }
        if (out_stack.empty()) in_to_out();
        int val = out_stack.top();
        out_stack.pop();
        return val;
    }
};

File: ctci/3_4.java
package ctci;

import java.util.Stack;

class MyQueue {
    private Stack<Integer> inStack;
    private Stack<Integer> outStack;

    public MyQueue() {
        inStack = new Stack<>();
        outStack = new Stack<>();
    }

    private void transfer() {
        while (!inStack.empty()) {
            outStack.push(inStack.pop());
        }
    }

    public boolean isEmpty() {
        return inStack.empty() && outStack.empty();
    }

    public void push(int x) {
        inStack.push(x);
    }

    public int peek() {
        if (isEmpty()) {
            throw new RuntimeException("Queue is empty!");
        }

        if (outStack.empty()) {
            transfer();
        }

        return outStack.peek();
    }

    public int pop() {
        if (isEmpty()) {
            throw new RuntimeException("Queue is empty!");
        }

        if (outStack.empty())
            transfer();

        return outStack.pop();
    }
}
